### PR TITLE
8256397: MultipleSelectionModel throws IndexOutOfBoundException

### PR DESCRIPTION
**Pull request by FlorianKirmaier**
Fixing IndexOutOfBoundsException in the MultipleSelectionModelBase and added a unit-test for it.
ticket: https://bugs.openjdk.java.net/browse/JDK-8256397
run test: `./gradlew --continue -PFULL_TEST=true controls:test --tests "*MultipleSelectionModelImplTest*"`

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8256397](https://bugs.openjdk.org/browse/JDK-8256397): MultipleSelectionModel throws IndexOutOfBoundException

### Reviewers
 * [Ajit Ghaisas](https://openjdk.org/census#aghaisas) (@aghaisas - **Reviewer**)
 * [Michael Strauß](https://openjdk.org/census#mstrauss) (@mstr2 - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx pull/353/head:pull/353` \
`$ git checkout pull/353`

Update a local copy of the PR: \
`$ git checkout pull/353` \
`$ git pull https://git.openjdk.org/jfx pull/353/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 353`

View PR using the GUI difftool: \
`$ git pr show -t 353`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/353.diff">https://git.openjdk.org/jfx/pull/353.diff</a>

</details>

### CODE REVIEW

**Comment by kleopatra** [comment]
good catch :)

No review, just a couple of comments:

- other subclasses of MultipleSelectionModelBase might have similar issues 
- the fix gets rid of the error (good!) but might fire incorrect changes (see below)
- there are quite a lot of open issues related to change notification, some might profit from a fix of this
- tests, tests, tests .. :)

selectIndices might involve disjoint intervals in the change (even for addition which is unusual in the "normal" implementations of observableLists)

       // initial selection, starting from empty
       selectionModel.selectIndices(0, /*1, IMPORTANT: remove some index */ 2,  3);
       System.out.println(change);
      // expected and actual change 
      { [0, 2, 3] added at 0 }

      // add selection of disjoint indices (assuming a longer items list)
       selectionModel.selectIndices(1, 5, 7);
       System.out.println(change);
      // actual
       { [1, 2, 3] added at 1 }
      // expected
     { [1] added at 1, [5, 7] added at 4 }

**Comment by FlorianKirmaier** [comment]
I've rebased the PR to the current master.
I've also added some more test-cases.
It would be great if we could finish the PR, because it's already a year old.

If I make the following call, I get the following changes:
```
selectionModel.selectIndices(0,2,3);
c.getAddedSubList() -> [0, 2, 3]
selectionModel.selectIndices(1,4,5);
c.getAddedSubList() -> [1, 2, 3]
```
This looks wrong to me, but I don't know the logic too well, can this be somehow correct?
On the other side, I would like to finish it, without looking for other bugs.

**Comment by christianheilmann** [comment]
How to proceed to get this PR reviewed?

**Comment by FlorianKirmaier** [comment]
I've just merged it with master
I hope someone will find time to review this.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/MultipleSelectionModelImplTest.java [hunk 1]
```diff
@@ -1445,4 +1446,31 @@ public void test_jdk_8088896() {
 
         assertEquals(2, model.getSelectedIndices().size());
     }
+
+
+    @Test
+    public void test_8256397() throws InterruptedException {
```
Comment (mstr2) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/MultipleSelectionModelImplTest.java:1452]: This test should probably get a meaningful name.

**Comment by mstr2** [comment]
I don't understand how the original code, as well as the proposed solution, can work at all. Let's take @kleopatra's example:
```java
selectionModel.selectIndices(0, 2, 3);
// ---> expected: { [0, 2, 3] added at 0 }
selectionModel.selectIndices(1, 5, 7);
// ---> expected: { [1] added at 1, [5, 7] added at 4 }
```
In order to find the two sub-ranges `[1]` and `[5, 7]`, we need to compare the _newly added_ indices to the _entire_ list of selected indices. Without doing that, we can't just "know" that there are two contiguous sub-ranges in `[1, 5, 7]`.

However, the algorithm that supposedly tries to do that, doesn't even look at the current list of selected indices (except for a single `indexOf` call, which doesn't do the trick):
```java
int pos = 0;
int start = 0;
int end = 0;

// starting from pos, we keep going until the value is
// not the next value
int startValue = sortedNewIndices.get(pos++);
start = indexOf(startValue);
end = start + 1;
int endValue = startValue;
while (pos < size) {
    int previousEndValue = endValue;
    endValue = sortedNewIndices.get(pos++);
    ++end;
    if (previousEndValue != (endValue - 1)) {
        _nextAdd(start, end);
        start = end;
        continue;
    }
    
    // special case for when we get to the point where the loop is about to end
    // and we have uncommitted changes to fire.
    if (pos == size) {
        _nextAdd(start, end);
    }
}
```
I think this algorithm mistakenly finds contiguous ranges within the list of _newly added_ indices, which is not what we want.

Here is an algorithm that works:
```java
int startIndex = indexOf(sortedNewIndices.get(0));
int endIndex = startIndex + 1;

for (int i = 1; i < sortedNewIndices.size(); ++i) {
    int currentValue = get(endIndex);
    int currentNewValue = sortedNewIndices.get(i);
    if (currentValue != currentNewValue) {
        _nextAdd(startIndex, endIndex);
        while (get(endIndex) != currentNewValue) ++endIndex;
        startIndex = endIndex++;
    } else {
        ++endIndex;
    }

    if (i == sortedNewIndices.size() - 1) {
        _nextAdd(startIndex, endIndex);
    }
}
```

Unfortunately, even with the working algorithm, the list change notifications are still not correct in all cases due to another (probably unrelated) bug in `SelectedIndicesList.get(int)`: If I remove the optimization that is implemented in that method, and replace it with a simple lookup, the notifications are always correct in my tests:
```java
@Override public Integer get(int index) {
    for (int lastGetIndex = 0, lastGetValue = bitset.nextSetBit(0);
         lastGetValue >= 0 || lastGetIndex == index;
         lastGetIndex++, lastGetValue = bitset.nextSetBit(lastGetValue + 1)) {
        if (lastGetIndex == index) {
            return lastGetValue;
        }
    }
    return -1;
}
```

Test 1:
```java
selectionModel.selectIndices(0, 2, 3);
// ---> { [0, 2, 3] added at 0 }
selectionModel.selectIndices(1, 5, 7);
// ---> { [1] added at 1, [5, 7] added at 4 }
```
Test 2:
```java
selectionModel.selectIndices(1, 3, 4);
// ---> { [1, 3, 4] added at 0 }
selectionModel.selectIndices(0, 5, 7);
// ---> { [0] added at 0, [5, 7] added at 4 }
selectionModel.selectIndices(6, 8, 9);
// ---> { [6] added at 5, [8, 9] added at 7 }
```

**Comment by FlorianKirmaier** [comment]
@mstr2 
Thank you for the corrected version of the algorithm, and also for highlighting further bugs.
I investigated the code a bit, and found the cause why the get method was so unreliable.

It turned out, that the optimization in the get method only works when the bit-set is not changed.
I've added some calls to the "reset" method, which resets the optimization when the bitset is changed.
I've added the correction into the PR.

I've also changed the tests, so we can easily test any case we come up with.
Also all the tests you've provided are now part of the PR and are green!

File: modules/javafx.controls/src/main/java/javafx/scene/control/MultipleSelectionModelBase.java [hunk 2]
```diff
@@ -827,6 +824,7 @@ public void clear(int index) {
             _beginChange();
             size = -1;
             bitset.clear(index);
+            if(index <= lastGetValue) reset();
```
Comment (aghaisas) [inline modules/javafx.controls/src/main/java/javafx/scene/control/MultipleSelectionModelBase.java:827]: There should be a space between `if` and `(`

File: modules/javafx.controls/src/main/java/javafx/scene/control/MultipleSelectionModelBase.java [hunk 3]
```diff
@@ -739,12 +741,14 @@ public void set(int index, int end, boolean isSet) {
...
             if (isSet) {
                 bitset.set(index, end, isSet);
+                if(index <= lastGetValue) reset();
                 int indicesIndex = indexOf(index);
                 int span = end - index;
                 _nextAdd(indicesIndex, indicesIndex + span);
             } else {
                 // TODO handle remove
                 bitset.set(index, end, isSet);
+                if(index <= lastGetValue) reset();
```
Comment (aghaisas) [inline modules/javafx.controls/src/main/java/javafx/scene/control/MultipleSelectionModelBase.java:751]: There should be a space between `if` and `(`

File: modules/javafx.controls/src/main/java/javafx/scene/control/MultipleSelectionModelBase.java [hunk 4]
```diff
@@ -739,12 +741,14 @@ public void set(int index, int end, boolean isSet) {
             size = -1;
             if (isSet) {
                 bitset.set(index, end, isSet);
+                if(index <= lastGetValue) reset();
```
Comment (aghaisas) [inline modules/javafx.controls/src/main/java/javafx/scene/control/MultipleSelectionModelBase.java:744]: There should be a space between `if` and `(`

File: modules/javafx.controls/src/main/java/javafx/scene/control/MultipleSelectionModelBase.java [hunk 5]
```diff
@@ -717,6 +718,7 @@ public void set(int index) {
             _beginChange();
             size = -1;
             bitset.set(index);
+            if(index <= lastGetValue) reset();
```
Comment (aghaisas) [inline modules/javafx.controls/src/main/java/javafx/scene/control/MultipleSelectionModelBase.java:721]: There should be a space between `if` and `(`

**Review by aghaisas — COMMENTED** [review]
The fix looks good.
I have identified minor spacing issues.

**Comment by aghaisas** [comment]
@mstr2, can you please review as well?

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/MultipleSelectionModelImplTest.java [hunk 6]
```diff
@@ -1445,4 +1446,65 @@ public void test_jdk_8088896() {
...
+                new int[][]{new int[]{6}, new int[]{8,9}});
+
+        testSelectedIndicesChangeEventsFactory(5, new int[]{6, 7}, 2, new int[]{1,0},
+                new int[][]{new int[]{0,1,2}});
+
+        testSelectedIndicesChangeEventsFactory(2, new int[]{3, 6,7}, 0, new int[]{1,4,5,8,9},
+                new int[][]{new int[]{0,1},new int[]{4,5},new int[]{8,9}});
+    }
+
+    public ListView creatListViewWithMultipleSelection() {
```
Comment (mstr2) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/MultipleSelectionModelImplTest.java:1470]: There's a typo in "creatList..."

**Review by mstr2 — COMMENTED** [review]
Looks good to me.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/MultipleSelectionModelImplTest.java [hunk 7]
```diff
@@ -1445,4 +1446,65 @@ public void test_jdk_8088896() {
...
+                new int[][]{new int[]{0}, new int[]{5,7}});
+
+        testSelectedIndicesChangeEventsFactory(0, new int[]{1, 3,4, 5, 7}, 6, new int[]{8,9},
+                new int[][]{new int[]{6}, new int[]{8,9}});
+
+        testSelectedIndicesChangeEventsFactory(5, new int[]{6, 7}, 2, new int[]{1,0},
+                new int[][]{new int[]{0,1,2}});
+
+        testSelectedIndicesChangeEventsFactory(2, new int[]{3, 6,7}, 0, new int[]{1,4,5,8,9},
+                new int[][]{new int[]{0,1},new int[]{4,5},new int[]{8,9}});
```
Comment (andy-goryachev-oracle) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/MultipleSelectionModelImplTest.java:1467]: I wonder if this test covers all the possible code paths in the implementation.  To illustrate just a subset of scenarios that should be tested, let's consider these scenarios (initial selection on the top line, change on the bottom line):

```
index 0 flipped, index 1 unchanged:
-I-
||-
index 1 unchanged
-|-
-|-
index 1 unchanged, index 2 flipped
-|-
-||
index 0 and 2 flipped
|--
--|
etc.
```


Perhaps it is possible to design a test that user random number generator to generate initial state and the changes, apply these to both MultipleSelectionModelBase and let's say a BitSet, and then compare the results - doing this 10,000 times (a sort of fuzzing test).  A model with 10 items might be sufficient.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/MultipleSelectionModelImplTest.java [hunk 8]
```diff
@@ -1445,4 +1446,65 @@ public void test_jdk_8088896() {
...
         assertEquals(2, model.getSelectedIndices().size());
     }
+
+
+    @Test
+    public void testSelectedIndicesChangeEvents() throws InterruptedException {
+
+        testSelectedIndicesChangeEventsFactory(0, new int[]{2,3}, 1, new int[]{5,7},
+                new int[][]{new int[]{1}, new int[]{5,7}});
+
```
Comment (andy-goryachev-oracle) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/MultipleSelectionModelImplTest.java:1456]: Do we have a test that tries to select indexes outside of the model?

**Comment by FlorianKirmaier** [comment]
First, thank you all for the code review and the suggestions.

I guess writing a TestFactory with randomly generated numbers would make sense. But it's also not trivial, because it requires reprogramming the logic of how the events are created, to ensure it's correct.

If someone finds time to do it, it's probably worth the time. Wouldn't be surprised if there is another error in this spaghetti code.

Because this PR is now 2 Years old ... I can tell you that it's tested for that time in a huge application with thousands of users.

This PR fixes an issue when editing entries in a ListView.

I don't really plan to extend the test further, unless someone points out an error, or suggests a simple modification.

**Review by mstr2 — APPROVED** [review]

**Review by aghaisas — APPROVED** [review]

**Pull request closed without merging**